### PR TITLE
feat: connect tool_calls to ToolOutputsBuffer in SoupTrainerCallback

### LINKED ISSUE #100: v0.44.1: trainer callback wiring for ToolOutputsBuffer
v0.44.0 ships `utils/tool_outputs.ToolOutputsBuffer` + `ToolCallTimer` but no trainer wires `record_call(...)` yet. For tool-calling SFT runs, the SoupTrainerCallback should observe `tool_calls` from each batch and emit a record to the buffer that the Web UI panel can render.

## Acceptance
- [ ] Plumbing in `monitoring/callback.py`.
- [ ] Web UI panel reads from the buffer.

### PR DESCRIPTION
**Pull request by Miriam-R-coder**
[#100] Add tool_calls tracking to SoupTrainerCallback

Captures tool execution metrics from HF logs and pipes them to the 
ToolOutputsBuffer for real-time monitoring.
Fixes #100

### CODE REVIEW

**Review by MakazhanAlpamys — COMMENTED** [review]
Welcome to Soup @Miriam-R-coder, and thanks for picking up #100! The direction is right -- threading a buffer through the callback is exactly what the issue asks for.

I want to flag a few things that need addressing before this can land. Don't take it as discouragement -- the wiring instinct is correct, the implementation just needs to match where tool calls actually flow through HF Trainer.

**1. record_call signature does not match**

The current call uses kwargs that don't exist:

```python
self.tool_output_buffer.record_call(
    step=state.global_step,
    tool_calls=tool_calls,
)
```

But `record_call` in soup_cli/utils/tool_outputs.py:62 is keyword-only and requires:

```python
def record_call(
    self, *,
    name: str,
    started_ts: float,
    duration_ms: float,
    success: bool,
    output_preview: str,
    error: Optional[str] = None,
) -> ToolCallRecord
```

So if this branch ever ran, it would raise `TypeError`. The buffer records one entry **per individual tool call** (with its own timing + output), not one entry per training step.

**2. on_log doesn't see tool_calls**

The on_log callback fires from `Trainer.log()` and only carries training metrics: `loss`, `learning_rate`, `grad_norm`, `train_steps_per_second`, plus whatever `compute_metrics` returns from eval steps. HF Trainer never puts `tool_calls` in that dict.

Tool calls live in:
- `model.generate()` output during inference / eval generation,
- a custom `compute_metrics` if the user wraps eval predictions,
- or a separate generation hook the user explicitly emits.

So `logs.get("tool_calls", None)` will always be None, and the buffer never gets fed. That makes the new branch effectively dead code -- which is why CI would still be green even with the API mismatch.

**3. Caller side isn't wired**

You added `tool_output_buffer` as a parameter to `SoupTrainerCallback.__init__`, but nothing in `commands/train.py` constructs a buffer and passes it in. So even if (1) and (2) were fixed, `self.tool_output_buffer` would always be `None` in practice. The wiring needs both ends.

**4. A test would catch the above**

A small unit test in `tests/` that:
- constructs the callback with a mock buffer,
- fires `on_log` with a logs dict containing a synthetic `tool_calls` payload,
- asserts `buffer.record_call(...)` was called with the expected kwargs,

would have flagged both (1) and (2) immediately. `tests/test_v0440_part_a.py` already has a buffer fixture pattern you can copy.

**Suggested direction**

The cleanest spot to record tool calls is **not** `on_log`. Two viable hooks:

- `on_evaluate` / a wrapped `compute_metrics`: scan the eval predictions for tool calls and record each one. This works for eval-time tool-call SFT.
- A fresh callback method invoked from a generation wrapper Soup controls: record per-call from the generator side, not the trainer side.

If you want, drop a note in #100 about which path you'd like to try and I can confirm the API shape before you write the next iteration. Happy to keep iterating -- the start is good, just needs to land in the right place.

Marking this as request-changes for now, but the door is fully open.

**Comment by MakazhanAlpamys** [comment]
Closing alongside #108. See the closing comment on #108 for the full rationale -- both PRs share the same issues (broken `soup train`, body that doesn't match diff, prior review feedback not addressed). Welcome to revisit #100 with a focused, tested implementation that addresses the original review.
